CORE-9052 beta1 to 5.0 2022.12.22 merge

vlajos opened the pull request:
CORE-9052
* Resolve input and reference input states in UtxoLedgerTransactionFactory.
* Which requires the UtxoLedgerPersistenceService. To avoid creating a circular dependency between SignedTxFactory and LedgerPersistenceService, we made the persistence service accessible via non-osgi. This needs to improved later.
* Removed some toLedgerTransaction() calls where the already available signed transaction has all the required information in the given context. (To spare the database lookups and adding @Suspendable)
* WrappedUtxoWireTransactionKryoSerializer
* ReferenceInputState -> ReferenceState

CORE-9014 
* Fixed issue with incorrect flow external event ID not set correctly for token claim release acknowledgements
* Fixed issue preventing flow resuming after failed token query.
* Fixed issue preventing token selection API flow injection.
* Removed old unused test code

CORE-9021
* Enable notary smoke test scenarios

CORE-9050
* Fix bug with Kryo serialization

API: https://github.com/corda/corda-api/pull/763

vlajos committed:
NOTICK notary selection first -> single (#2803)

Use safer single() instead of first() in notary selection logic.

vlajos committed:
CORE-9051 Fix FilteredTransactions for empty output's sizeproofs (#2797)

Co-authored-by: Christian Sailer <christian.sailer@r3.com>

mbrkic-r3 committed:
CORE-9055 Creating new entity manager in each function of UtxoPersistenceServiceImpl (#2808)

vlajos committed:
CORE-9052 ledger tx inputs (#2799)

* Resolve input and reference input states in UtxoLedgerTransactionFactory.
* Which requires the UtxoLedgerPersistenceService. To avoid creating a circular dependency between SignedTxFactory and LedgerPersistenceService, we made the persistence service accessible via non-osgi. This needs to be improved later.
* Removed some toLedgerTransaction() calls where the already available signed transaction has all the required information in the given context. (To spare the database lookups and adding @Suspendable)
* WrappedUtxoWireTransactionKryoSerializer
* ReferenceInputState -> ReferenceState

vlajos committed:
Merge branch 'release/os/5.0-Beta1' into vlajos/5.0-Beta1-to-5.0-2022.12.22-merge

Commit:
CORE-8922: bump postgres driver (#2816)

bump minor version

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceServiceImpl.kt:
```diff
@@ -128,8 +128,9 @@ class UtxoLedgerPersistenceServiceImpl @Activate constructor(
-        return utxoSignedTransactionFactory.create(wireTransaction, signatures)
+    private fun SignedTransactionContainer.toSignedTransaction(utxoLedgerPersistenceService: UtxoLedgerPersistenceService)
+    : UtxoSignedTransaction {
+        return utxoSignedTransactionFactory.create(wireTransaction, signatures, utxoLedgerPersistenceService)
```
```suggestion
        return utxoSignedTransactionFactory.create(wireTransaction, signatures, this)
```

Seems a little odd to take `utxoLedgerPersistenceService` as an argument, when it is `this` object and this function is private? Suggested approach results in being able to revert the change to line 62 as well.

vlajos replied:
Updated here: https://github.com/corda/corda-runtime-os/pull/2819/files#diff-6ca8a2c0e6178df692e3bc6b4d24858f329a12a609ffa359752d20bc07971377

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt:
```diff
@@ -103,15 +104,15 @@ data class UtxoSignedTransactionImpl(
+        if (signatories.any {
                 !it.isFulfilledBy(appliedSignatories) // isFulfilledBy() helps to make this working with CompositeKeys.
-            }){
+            }) {
             throw TransactionVerificationException(id, "There are missing signatures", null)
         }
```
Something looks off with the indentation here (think this was also the case before. Would suggest moving the opening brace to the next line., or adding another layer of indentation for the throw statement.

Also, the error may be more useful if it said which expected signatures are missing?

vlajos replied:
Updated here: https://github.com/corda/corda-runtime-os/pull/2819/files#diff-ba66f90f1d378b1fc768f02d19f5ec2b6ed7c1c6b947c3e0cc5710f8a15c4afb
The logging side has already a ticket: https://r3-cev.atlassian.net/browse/CORE-8935

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt:
```diff
@@ -161,7 +163,7 @@ data class UtxoTransactionBuilderImpl(
                 && other.attachments == attachments
                 && other.commands == commands
                 && other.inputStateRefs == inputStateRefs
-                && other.referenceInputStateRefs == referenceInputStateRefs
+                && other.referenceStateRefs == referenceStateRefs
```
Not new to this change - but does this imply that for two of these to be equal, the exact order of these must be the same for two objects to be equal? I.e. if I had two objects that were the same other than one defined input states as [A,B] and another defined them as [B,A], should these be considered equal, or different?

vlajos replied:
AFAIK if the order of inputs is different the two transactions are supposed to be treated as different ones.
Their merkle trees and txids will be different without sorting them anyway.

blsemo replied:
Technically, the order doesn't matter - but I don't think it's a big issue. I would normally only expect transaction builders to be equal if they are copies of the same builder, in which case the order is the same. 
I can't really see a use case where someone would create two equal transation builders completely independently.
One might even argue that the order of states in these lists matters, as re-ordering them changes the Merkle tree of the SignedTransaction that will be created.

relyafi replied:
I don't have a strong opinion either way, just wanted to ensure we'd considered what our expected / correct behavior is.

All I'd say is that we never know how customers will use our objects, so we should be prepared to justify our behavior either way.

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/UtxoLedgerTransactionFactory.kt:
```diff
@@ -0,0 +1,12 @@
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction
+
+interface UtxoLedgerTransactionFactory {
```
Think this needs a kdoc comment.

vlajos replied:
I have added few lines here: https://github.com/corda/corda-runtime-os/pull/2819/files#diff-4ee14d284a0c0ef2c15f2c0760c832534c7662f5cef6c4a5eeac36f786959096
(It is not in the public API.)

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt:
```diff
@@ -0,0 +1,60 @@
+        wireTransaction: WireTransaction
+    ): UtxoLedgerTransaction {
+        val wrappedUtxoWireTransaction = WrappedUtxoWireTransaction(wireTransaction, serializationService)
+        val sourceTransactions =
+            (wrappedUtxoWireTransaction.inputStateRefs + wrappedUtxoWireTransaction.referenceStateRefs)
+                .map { it.transactionHash }
+                .toSet()
+                .associateWith { it ->
+                    utxoLedgerPersistenceService.find(it)?.outputStateAndRefs
+                        ?: throw (CordaRuntimeException("Could not find transaction $it when fetching input states."))
+                }
+        val inputStateAndRefs =
+            wrappedUtxoWireTransaction.inputStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Input state not found ${it.transactionHash} ${it.index}"))
+            }
+
+        val referenceStateAndRefs =
+            wrappedUtxoWireTransaction.referenceStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Reference state not found ${it.transactionHash} ${it.index}"))
+            }
```
This is probably the most elegant solution, but also looks like it's doing an unnecessary additional iteration of the input and reference states. Might be more efficient to do one loop each of these and to setup the source transactions and relevant stateandrefs object in one loop.

blsemo replied:
I am not so sure about this one - the expensive bit here is finding the source transaction, not looping through the inputs/references. 
So either we need one iteration caching all txs temporarily, or this two phase approach. A comment explaining the reasoning might be useful, though.

relyafi commented on components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/factory/impl/UtxoLedgerTransactionFactoryImpl.kt:
```diff
@@ -0,0 +1,60 @@
+        val inputStateAndRefs =
+            wrappedUtxoWireTransaction.inputStateRefs.map { it ->
+                sourceTransactions[it.transactionHash]?.get(it.index)
+                    ?: throw (CordaRuntimeException("Input state not found ${it.transactionHash} ${it.index}"))
```
```suggestion
                    ?: throw (CordaRuntimeException("Input state not found ${it.ref}"))
```

Should use the standard `toString()` defined on the reference part, which will add a colon between the hash and index and ensures formatting consistency.

Comment also applies to line 51 as well.

vlajos replied:
Thanks. I have updated them:
https://github.com/corda/corda-runtime-os/pull/2819/files#diff-e512a249344bb1ddd2949e337a93c0e25ef4c25b2af08a23a643bc1715c04122

relyafi reviewed (commented):
Notary bits LGTM, have some general comments.

Appreciate this is a merge PR so might need some thought as to how and where to address these comments if you decide to fix.

nkovacsx committed:
CORE-9050 Fix bug with Kryo serialization (#2817)

vlajos commented:
> 



> Appreciate this is a merge PR so might need some thought as to how and where to address these comments if you decide to fix.

Thanks. I am addressing them here: https://github.com/corda/corda-runtime-os/pull/2819

owenstanford committed:
CORE-9014 (#2788)

- Fixed issue with incorrect flow external event ID not set correctly for token claim release acknowledgements
- Fixed issue preventing flow resuming after failed token query.
- Fixed issue preventing token selection API flow injection.
- Removed old unused test code

vlajos committed:
CORE-9052 Fix up minor details

Minor changes
Flesh out kDocs

Co-authored-by: Christian Sailer <christian.sailer@r3.com>

nkovacsx committed:
CORE-9021 Enable notary smoke test scenarios (#2809)

* CORE-9021 Enable smoke test scenarios, bug fixes

* Re-add scenarios for notary smoke tests

* Revert version

* Change non-validating notary test flow logic

* use unknownTxId

vlajos committed:
Merge branch 'release/os/5.0-Beta1' into vlajos/5.0-Beta1-to-5.0-2022.12.22-merge

vlajos committed:
CORE-9052 Api version

vlajos committed:
CORE-9052 Api version
